### PR TITLE
ENH: Added volume source support for fio

### LINKED ISSUE #88: Fio: hostpath support
hostpath parameter is helpful to measure disk speed on the nodes. Fio should be extended to enhance our flexibility.

Further read: https://kubernetes.io/docs/concepts/storage/volumes/#hostpath

### LINKED ISSUE #89: Fio: emptydir support
Emptydir is often used in Kubernetes deployments. Supporting it in fio will provide valuable insights to the users about it's performance. 

Note: this could be worked around with hostdir support (#88), but from the usability perspective it is easier to deal with emptydirs.

### PR DESCRIPTION
**Pull request by julio-santiesteban**
Fixes #88
Fixes #89

### CODE REVIEW

File: api/v1alpha1/fio_types.go [hunk 1]
```diff
@@ -57,6 +61,11 @@ type FioSpec struct {
 	// PodConfig contains the configuration for the benchmark pod, including
 	// pod labels and scheduling policies (affinity, toleration, node selector...)
 	PodConfig PodConfigurationSpec `json:"podConfig,omitempty"`
+
+	// Volume contains the configuration for the volume that the fio job should
+	// run on. If missing, no volume will attached to the job and Docker's layered
+	// fs performance will be measured
+	Volume *FioVolumeSpec `json:"volume,omitempty"`
```
Comment (wgershwitz) [inline api/v1alpha1/fio_types.go]: Optional flag is missing. I'm not quite sure about how helpful it is to measure docker's layered fs performance. are we sure we want to keep this?

File: api/v1alpha1/fio_types.go [hunk 2]
```diff
@@ -57,6 +61,11 @@ type FioSpec struct {
 	// PodConfig contains the configuration for the benchmark pod, including
 	// pod labels and scheduling policies (affinity, toleration, node selector...)
 	PodConfig PodConfigurationSpec `json:"podConfig,omitempty"`
```
Comment (wgershwitz) [inline api/v1alpha1/fio_types.go]: unrelated, but optional flag is missing here.

File: controllers/fio/job.go [hunk 3]
```diff
@@ -62,14 +62,15 @@ func NewJob(cr *perfv1alpha1.Fio, configMap *corev1.ConfigMap, pvcName *string)
 			Name: "custom-jobs", MountPath: "/custom-jobs",
 		})
 	}
-	if pvcName != nil {
+	if cr.Spec.Volume != nil {
+		volumeSource := corev1.VolumeSource{}
```
Comment (wgershwitz) [inline controllers/fio/job.go]: I think we need to add emptyDir as volumeSource. At least in yaml that's what you'd do:
https://kubernetes.io/docs/concepts/storage/volumes/

**Comment by wgershwitz** [comment]
Given that we would like to wire out emptyDir and hostPath, it might make sense to expose the whole 'volumes' section of pod. That'd solve two tickets at once. WDYT?

File: controllers/fio/job.go (hunk 3 above)
Comment (julio-santiesteban) [reply controllers/fio/job.go]: No, I tested it, the above code creates an emptydir.

File: api/v1alpha1/fio_types.go (hunk 2 above)
Comment (julio-santiesteban) [reply api/v1alpha1/fio_types.go]: resolved by ed1526156

File: api/v1alpha1/fio_types.go (hunk 1 above)
Comment (otto-von-bivouac) [reply api/v1alpha1/fio_types.go]: @wgershwitz is still valid here I think

File: docs/benchmarks/fio.md [hunk 4]
```diff
@@ -13,10 +13,10 @@ Kubestone generates a Kubernetes Job from each fio CR that will run a single pod
 
 When `customJobFiles` are specified in the CR a ConfigMap will be created to hold the content of the job files. The entries in the ConfigMap named using the following pattern: `customJobN`, where N is the item in the customJobFiles list.
 
-If `PersistentVolumeClaim` is provided in the CR the respective PVC will be used for benchmarking.
+If `Volume` is specified in the CR, a volume will be mounted to the pod and the fio job will run inside that volume. If inside `Volume` `PersistentVolumeClaimName` is provided, the respective PVC will be used. If `PersistentVolumeClaim` is provided with a complete PVC definition (size, storage class, etc.), the described PVC will be created and used by the benchmark.
```
Comment (otto-von-bivouac) [inline docs/benchmarks/fio.md]: s/fio job will run inside/fio job will use/
s/If inside `Volume` `PersistentVolumeClaimName`/If `Volume.PersistentVolumeClaimName`/

File: api/v1alpha1/fio_types.go [hunk 5]
```diff
@@ -17,26 +17,28 @@ limitations under the License.
...
-	Image ImageSpec `json:"image"`
-
-	// PersistentVolumeClaimName is the name of the existing persistence volume
-	// claim that will be used by the benchmark pod. If undefined, you can either
-	// use PersistentVolumeClaim to create and use a PVC, or nothing to run the
-	// benchmark inside the pod.
-	// +optional
-	PersistentVolumeClaimName *string `json:"persistentVolumeClaimName,omitempty"`
+// FioVolumeSpec contains the configuration for the volume that the fio job
+// should run on
```
Comment (otto-von-bivouac) [inline api/v1alpha1/fio_types.go]: should use for benchmarking

File: config/samples/fio/overlays/pvc/patch.yaml [hunk 6]
```diff
@@ -3,7 +3,11 @@ kind: Fio
...
-  persistentVolumeClaim:
-    accessModes:
-    - ReadWriteOnce
-    size: 1Gi
+  volume:
+    # persistentVolumeClaim:
+    #   accessModes:
+    #   - ReadWriteOnce
+    #   size: 1Gi
+    volumeSource:
```
Comment (otto-von-bivouac) [inline config/samples/fio/overlays/pvc/patch.yaml]: Can we please have the emptyDir as the default example here?

**Comment by julio-santiesteban** [comment]
> Since the pvc kustomization is changed, the quickstart guide should also be updated. Thank you!

Done (295919c).

File: api/v1alpha1/fio_types.go (hunk 5 above)
Comment (julio-santiesteban) [reply api/v1alpha1/fio_types.go]: done

File: api/v1alpha1/fio_types.go (hunk 1 above)
Comment (julio-santiesteban) [reply api/v1alpha1/fio_types.go]: done

File: docs/benchmarks/fio.md (hunk 4 above)
Comment (julio-santiesteban) [reply docs/benchmarks/fio.md]: done

**Pull request merged**